sel4test-sim: enable MCS simulation tests

lsf37 opened the pull request:
It appears that switching on MCS for simulation runs "just works".

It'll increase test time for simulation by about a factor of 2, but I think that's worth it and might reduce hardware test load. In particular, it will actually compile MCS variants of all architectures without us having to request a more expensive `hw-build`.

lsf37 commented:
Still not fully finished with testing, but so far `armv7a` and `riscv` seem happy enough.

lsf37 commented:
Adding another run on CI [here](https://github.com/lsf37/seL4/actions/runs/3441178835/jobs/5740613275) just to be sure before I merge.

lsf37 commented:
With the failing parts (x86, zynq7000) removed, this now appears to be stable and working.

Overall runtime is increased from 9min on the longest job to 13min. If we find 13min too annoying, we can split out the build matrix a bit more, probably best into separate runs for zynq and sabre, but I'm going to merge it first as is.

lsf37 commented:
Unfortunately, after about 10 runs, I have to conclude it's not as stable as I thought.

Out of 8 runs, I have 3 where `ARMVIRT_release_clang_MCS_64` fails somewhere, and 2 where `SPIKE32_debug_gcc_MCS_32` fails. I have not seen failures of other jobs yet, but it does sound like host timing difference do influence success/failure of the simulation tests under MCS.

lsf37 commented:
Potentially it's just that the kernel WCET is set insufficiently high when run under emulation in qemu. 

@kent-mcleod or @axel-h do you have any experience with that?

Or it is just unrealistic to want to run MCS under emulation -- on the other hand it does seem to work most of the time, so maybe there is hope.

In any case, I'll reset this to draft so we don't merge it accidentally.

kent-mcleod commented:
> @kent-mcleod or @axel-h do you have any experience with that?

Because the qemu emulation isn't cycle accurate, the clock can be observed to skip ahead in large chunks based on scheduling of the host machine. One way that I've seen that seems to work around this slightly is by taking multiple sample points and discarding outliers. EG, if you want to implement a 50ms delay, you break it up into 10 5ms delays, and so any big jumps can only affect 10% of the total blocking time.  I don't have a good way to apply this to the core time accounting in MCS, but it may be possible to use it to write more robust tests.

lsf37 commented:
I'll close this for now -- if we do want to have MCS simulation tests, we'll have to do something more sophisticated.

midnightveil commented on sel4test-sim/build.py:
```diff
@@ -18,21 +18,35 @@
+
+    # MCS simulation does not work yet for x86, fails SCHED0011 and INTERRUPT0005
+    if plat.arch == 'x86' and build.is_mcs():
+        return False
```
Disabled in sel4test via PR 156, so I think we should be good to go enabling this again.

midnightveil commented on sel4test-sim/build.py:
```diff
@@ -18,21 +18,35 @@
+
+    # MCS simulation is hanging on qemu for ZYNQ7000
+    if plat.name == 'ZYNQ7000' and build.is_mcs():
+        return False
```
I see the following locally when I run with my stress trick. Without it it works fine.

```
Test CANCEL_BADGED_SENDS_0001 passed
Starting test 22: CANCEL_BADGED_SENDS_0002
seL4 failed assertion 'getCurrentTime() < deadline || globalTimer->isr == 1u' at /home/julia/ts/sel4test-manifest-qemu-testing/kernel/include/drivers/timer/arm_global.h:76 in function setDeadline
halting...
Kernel entry via Syscall, number: 1, Call
Cap type: 2, Invocation tag: 1
```

midnightveil replied:
Sometimes it is instead:

```
Running test CANCEL_BADGED_SENDS_0002 (cancelBadgedSends deletes caps)
<<seL4(CPU 0) [decodeInvocation/643 T0xe000ea00 "child of: '35'" @1d918]: Attempted to invoke a null cap #169.>>
<<seL4(CPU 0) [decodeInvocation/643 T0xe000ea00 "child of: '35'" @1d918]: Attempted to invoke a null cap #169.>>
<<seL4(CPU 0) [decodeInvocation/643 T0xe000ee00 "child of: '35'" @1d918]: Attempted to invoke a null cap #287.>>
<<seL4(CPU 0) [decodeInvocation/643 T0xe000ea00 "child of: '35'" @1d918]: Attempted to invoke a null cap #169.>>
<<seL4(CPU 0) [decodeInvocation/643 T0xe000ee00 "child of: '35'" @1d918]: Attempted to invoke a null cap #287.>>
<<seL4(CPU 0) [decodeInvocation/643 T0xe04c4200 "child of: '35'" @1d918]: Attempted to invoke a null cap #341.>>
<<seL4(CPU 0) [decodeInvocation/643 T0xe000ea00 "child of: '35'" @1d918]: Attempted to invoke a null cap #169.>>
<<seL4(CPU 0) [decodeInvocation/643 T0xe000ee00 "child of: '35'" @1d918]: Attempted to invoke a null cap #287.>>
<<seL4(CPU 0) [decodeInvocation/643 T0xe04c4200 "child of: '35'" @1d918]: Attempted to invoke a null cap #341.>>
<<seL4(CPU 0) [decodeInvocation/643 T0xe04c4600 "child of: '35'" @1d918]: Attempted to invoke a null cap #405.>>
<<seL4(CPU 0) [decodeInvocation/643 T0xe000ea00 "child of: '35'" @1d918]: Attempted to invoke a null cap #169.>>
<<seL4(CPU 0) [decodeInvocation/643 T0xe000ee00 "child of: '35'" @1d918]: Attempted to invoke a null cap #287.>>
<<seL4(CPU 0) [decodeInvocation/643 T0xe04c4200 "child of: '35'" @1d918]: Attempted to invoke a null cap #341.>>
<<seL4(CPU 0) [decodeInvocation/643 T0xe04c4600 "child of: '35'" @1d918]: Attempted to invoke a null cap #405.>>
<<seL4(CPU 0) [decodeInvocation/643 T0xe04c4e00 "child of: '35'" @1d918]: Attempted to invoke a null cap #454.>>
<<seL4(CPU 0) [decodeInvocation/643 T0xe000ea00 "child of: '35'" @1d918]: Attempted to invoke a null cap #169.>>
<<seL4(CPU 0) [decodeInvocation/643 T0xe000ee00 "child of: '35'" @1d918]: Attempted to invoke a null cap #287.>>
<<seL4(CPU 0) [decodeInvocation/643 T0xe04c4200 "child of: '35'" @1d918]: Attempted to invoke a null cap #341.>>
<<seL4(CPU 0) [decodeInvocation/643 T0xe04c4600 "child of: '35'" @1d918]: Attempted to invoke a null cap #405.>>
<<seL4(CPU 0) [decodeInvocation/643 T0xe04c4e00 "child of: '35'" @1d918]: Attempted to invoke a null cap #454.>>
<<seL4(CPU 0) [decodeInvocation/643 T0xe04eb600 "child of: '35'" @1d918]: Attempted to invoke a null cap #568.>>
<<seL4(CPU 0) [decodeInvocation/643 T0xe000ea00 "child of: '35'" @1d918]: Attempted to invoke a null cap #169.>>
<<seL4(CPU 0) [decodeInvocation/643 T0xe000ee00 "child of: '35'" @1d918]: Attempted to invoke a null cap #287.>>
<<seL4(CPU 0) [decodeInvocation/643 T0xe04c4200 "child of: '35'" @1d918]: Attempted to invoke a null cap #341.>>
<<seL4(CPU 0) [decodeInvocation/643 T0xe04c4600 "child of: '35'" @1d918]: Attempted to invoke a null cap #405.>>
<<seL4(CPU 0) [decodeInvocation/643 T0xe04c4e00 "child of: '35'" @1d918]: Attempted to invoke a null cap #454.>>
<<seL4(CPU 0) [decodeInvocation/643 T0xe04eb600 "child of: '35'" @1d918]: Attempted to invoke a null cap #568.>>
<<seL4(CPU 0) [decodeInvocation/643 T0xe04eba00 "child of: '35'" @1d918]: Attempted to invoke a null cap #628.>>
<<seL4(CPU 0) [decodeInvocation/643 T0xe000ea00 "child of: '35'" @1d918]: Attempted to invoke a null cap #169.>>
<<seL4(CPU 0) [decodeInvocation/643 T0xe000ee00 "child of: '35'" @1d918]: Attempted to invoke a null cap #287.>>
<<seL4(CPU 0) [decodeInvocation/643 T0xe04c4200 "child of: '35'" @1d918]: Attempted to invoke a null cap #341.>>
<<seL4(CPU 0) [decodeInvocation/643 T0xe04c4600 "child of: '35'" @1d918]: Attempted to invoke a null cap #405.>>
seL4 failed assertion 'getCurrentTime() < deadline || globalTimer->isr == 1u' at /home/julia/ts/sel4test-manifest-qe
halting...
Kernel entry via Interrupt, irq 27
```

```
vka_alloc_object_at_maybe_dev@object.h:57 Failed to allocate object of size 2147483648, error 1
vka_alloc_object_at_maybe_dev@object.h:57 Failed to allocate object of size 1073741824, error 1
vka_alloc_object_at_maybe_dev@object.h:57 Failed to allocate object of size 536870912, error 1
vka_alloc_object_at_maybe_dev@object.h:57 Failed to allocate object of size 268435456, error 1
seL4 failed assertion 'getCurrentTime() < deadline || globalTimer->isr == 1u' at /home/julia/ts/sel4test-manifest-qe
halting...
Kernel entry via Syscall, number: 1, Call
Cap type: 2, Invocation tag: 1
```

(Note: it's not to do with the badged send test, it happens both before and after it).

This doesn't seem to happen for the generic qemu-arm-virt aarch32 platform, though.

midnightveil replied:
The QEMU platform not having this issue would be because it doesn't use the global timer.

Gerwin added these asserts in 2023, which is after this PR, so I expect that the hang has become an assertion failure. At a guess, this is not guaranteed by QEMU... (EDIT: looks like it should be, a write will immediately update status if the comparison is true). ~~That, or we experience a comparator overflow before writing the high bits. (EDIT: That's nonsense, it's the comparator not the counter).~~


```c
/** DONT_TRANSLATE */
static inline void setDeadline(ticks_t deadline)
{
    /* disable cmp */
    globalTimer->control &= ~(BIT(COMP_ENABLE));
    /* set low bits */
    globalTimer->comparatorLower = (uint32_t) deadline;
    /* set high bits */
    globalTimer->comparatorUpper = (uint32_t)(deadline >> 32llu);
    /* enable cmp */
    globalTimer->control |= BIT(COMP_ENABLE);
    /* Assert that either the deadline is in the future or that the IRQ has already been raised.
       This should be guaranteed by hardware from r2p0 on of the Cortex-A9 MPCore Technical
       Reference Manual (r2p0 published in 2009) */
    assert(getCurrentTime() < deadline || globalTimer->isr == 1u);
}
```

midnightveil replied:
Adding:

```c
    ticks_t current = getCurrentTime();
    uint32_t isr;
    if (!(current < deadline || (isr = globalTimer->isr) == 1u)) {
        printf("current: 0x%llx, deadline: 0x%llx, isr: %x\n", current, deadline, isr);
        assert(current < deadline || isr == 1u);
    }
```

I get, which makes no sense to me based on the QEMU source, unless we've not enabled the timer at all.

```
current: 0xcd88f58f, deadline: 0xcd7ed152, isr: 0
seL4 failed assertion 'current < deadline || isr == 1u' at /home/julia/ts/sel4test-manifest-qemu-testing/kernel/ince
halting...
Kernel entry via Syscall, number: 1, Call
Cap type: 4, Invocation tag: 0

```

midnightveil replied:
tldr; I think we should disable any platforms that use the global timer.

lsf37 replied:
Oh, yes, of course, that makes perfect sense. MCS doesn't work for those.

midnightveil replied:
Oh, MCS is unsupported on ARM32?

lsf37 replied:
Maybe I'm confusing something. I thought I saw comments somewhere in the code that platforms with just the global timer don't work for MCS. So either most platforms have other timers that MCS uses or it's just that they don't work with SMP + MCS.

midnightveil replied:
There is the `KernelPlatformSupportsMCS` in allwinnerA20 and APQ8024; which use the... arm_generic? timer. (That is what modern AArch64 platforms use ??? and clearly supports MCS?).

Indanz replied:
> I see the following locally when I run with my stress trick. Without it it works fine.
> 
> ```
> Test CANCEL_BADGED_SENDS_0001 passed
> Starting test 22: CANCEL_BADGED_SENDS_0002
> seL4 failed assertion 'getCurrentTime() < deadline || globalTimer->isr == 1u' at /home/julia/ts/sel4test-manifest-qemu-testing/kernel/include/drivers/timer/arm_global.h:76 in function setDeadline
> ```

Does this keep happening if you mark `globalTimer` volatile? Currently the compiler is free to re-order those accesses it seems...

midnightveil replied:
It already is marked as volatile (the pointer isn't but the pointer target timer_t type is a volatile struct).

Indanz replied:
Ah, missed that. No idea then.

~~Maybe we overflow twice, and the if-check isn't enough to catch it? IIRC, the QEMU timer frequency was insanely high.~~ Edit: No, can't be that either, it's a 2x32-bit timer.

midnightveil committed:
sel4test-sim: enable MCS simulation tests

As we've disabled the sel4tests that are simulation-sensitive
in sel4test, we can now enable MCS more generally so that
we have a (somewhat-limited) simulation check on our PRs for
basic functionality and compilation.

The ZYNQ7000 causes seL4 assertions when the host system is
under high load, so we disable that now.

Signed-off-by: Julia Vassiliki <julia.vassiliki@unsw.edu.au>

lsf37 reviewed (commented):
Happy for you to merge. (I can't approve because I opened the PR :-))
